core/src/main/java/org/elasticsearch/cockroach/TransportCockroachAction.java

```diff
@@ -0,0 +1,115 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cockroach;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+/**
+ * An action that can survive restart of requesting or executing node.
+ * These actions are using cluster state rather than only transport service to send requests and responses.
+ */
+public abstract class TransportCockroachAction<Request extends CockroachRequest<Request>, Response extends CockroachResponse>
+    extends HandledTransportAction<Request, Response> {
+
+    private final CockroachService cockroachService;
+
+    private final String executor;
+
+    protected TransportCockroachAction(Settings settings, String actionName, boolean canTripCircuitBreaker, ThreadPool threadPool,
+                                       TransportService transportService, CockroachService cockroachService, ActionFilters actionFilters,
+                                       IndexNameExpressionResolver indexNameExpressionResolver,
+                                       Supplier<Request> requestSupplier, Supplier<Response> responseSupplier, String executor) {
+        super(settings, actionName, canTripCircuitBreaker, threadPool, transportService, actionFilters, indexNameExpressionResolver,
+            requestSupplier);
+        this.cockroachService = cockroachService;
+        this.executor = executor;
+        cockroachService.registerRequestHandler(actionName, this, requestSupplier, responseSupplier, executor);
+    }
+
+    /**
+     * Returns the node id where the request has to be executed
+     */
+    public abstract DiscoveryNode executorNode(Request request, ClusterState clusterState);
+
+    /**
+     * Returns the node id where the response should be processed
+     */
+    public abstract DiscoveryNode responseNode(Request request, ClusterState clusterState);
```

Comment:
I'm wondering if we should either change the name of `responseNode` or keep the same name but make the comments more descriptive as to the exact role of the response node, as it is the place where the results of the task are processed and finalization occurs, but not necessarily where a response might be sent to the client..